text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,356 @@
+consistently formatted and flexible error reports. Today, there are 2 such
+forms of context that are traditionally gathered, `backtrace` and `source`.
+
+However, the current approach of promoting each form of context to a fn on the
+`Error` trait doesn't leave room for forms of context that are not commonly
+used, or forms of context that are defined outside of the standard library.
+
+## Example use cases this enables
+
+* using `backtrace::Backtrace` instead of `std::backtrace::Backtrace`
+* zig-like Error Return Traces by extracting `Location` types from errors
+  gathered via `#[track_caller]` or some similar mechanism.
+* error source trees instead of chains by accessing the source of an error as a
+  slice of errors rather than as a single error, such as a set of errors caused
+  when parsing a file
+* `SpanTrace` a backtrace like type from the `tracing-error` library
+* Help text such as suggestions or warnings attached to an error report
+
+By adding a generic form of these functions that works around the restriction
+on generics in vtables we could support a greater diversity of error handling
+needs and make room for experimentation with new forms of context in error
+reports.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Error handling in rust consists mainly of two steps, creation/propogation and
+reporting. The `std::error::Error` trait exists to bridge the gap between these
+two steps. It does so by acting as a consistent interface that all error types
+can implement to allow error reporting types to handle them in a consistent
+manner when constructing reports for end users.
+
+The error trait accomplishes this by providing a set of methods for accessing
+members of `dyn Error` trait objects. The main member, the error message
+itself, is handled by the `Display` trait which is a requirement for
+implementing the Error trait. For accessing `dyn Error` members it provides the
+`source` function, which conventionally represents the lower level error that
+caused the current error. And for accessing a `Backtrace` of the state of the
+stack when an error was created it provides the `backtrace` function. For all
+other forms of context relevant to an Error Report the error trait provides the
+`context`/`provide_context` functions.
+
+As an example of how to use these types to construct an error report lets
+explore how one could implement an error reporting type that retrieves the
+Location where each error in the chain was created, if it exists, and renders
+it as part of the chain of errors. Our end goal is to get an error report that
+looks something like this:
+
+```
+Error:
+    0: Failed to read instrs from ./path/to/instrs.json
+        at instrs.rs:42
+    1: No such file or directory (os error 2)
+```
+
+The first step is to define or use a Location type. In this example we will
+define our own but we could use also use `std::panic::Location` for example.
```

Comment:
Ah, I might be the only person in the world who finds this really confusing but this really tripped me up (I've been staring at core::panic::Location for months)

Reply:
Specifically what tripped me up was ending this sentence referencing std but then still defining our own. I'd pick one or the other and run with it.